src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,419 @@
+                    putMutex(t.getClass());
+                    throw t;
+                }
+            } finally {
+                supplying = false;
+            }
+        }
+    }
+
+    private static final Function<Object, String> ERROR_MESSAGE_EXTRACTOR = new Function<Object, String>() {
+        @Override
+        public String apply(Object stableValue) {
+            StableValueImpl<?> svi = (StableValueImpl<?>) stableValue;
+            return ((Class<?>) svi.acquireMutex())
+                    .getName();
+        }
+    };
+
+    @Override
+    public String toString() {
+        return StableUtil.toString(this, ERROR_MESSAGE_EXTRACTOR);
+    }
+
+    @SuppressWarnings("unchecked")
+    private V valueVolatile() {
+        return (V)UNSAFE.getReferenceVolatile(this, VALUE_OFFSET);
+    }
+
+    private void setValue(V value) {
+        if (value != null) {
+            putValue(value);
+        }
+        // Crucially, indicate a value is set _after_ it has actually been set.
+        putState(value == null ? NULL : NON_NULL);
+        putMutex(TOMBSTONE); // We do not need a mutex anymore
+    }
+
+    private void putValue(V value) {
+        // This prevents partially initialized objects to be observed
+        // under normal memory semantics.
+        freeze();
+        UNSAFE.putReferenceVolatile(this, VALUE_OFFSET, value);
+    }
+
+    private int stateVolatile() {
+        return UNSAFE.getIntVolatile(this, STATE_OFFSET);
+    }
+
+    private void putState(int newValue) {
+        // This prevents `this.value` to be seen before `this.state` is seen
+        freeze();
+        UNSAFE.putIntVolatile(this, STATE_OFFSET, newValue);
+    }
+
+    private Object acquireMutex() {
+        Object mutex = UNSAFE.getReferenceVolatile(this, MUTEX_OFFSET);
+        if (mutex == null) {
+            mutex = caeMutex();
+        }
+        return mutex;
+    }
+
+    private Object caeMutex() {
+        final var created = new Object();
+        final var witness = UNSAFE.compareAndExchangeReference(this, MUTEX_OFFSET, null, created);
+        return witness == null ? created : witness;
+    }
+
+    private void putMutex(Object value) {
+        UNSAFE.putReferenceVolatile(this, MUTEX_OFFSET, value);
+    }
+
+    // Factories
+
+    public static <V> StableValueImpl<V> of() {
+        return new StableValueImpl<>();
+    }
+
+    public static <V> StableValue<V> ofBackground(ThreadFactory threadFactory,
+                                                  Supplier<? extends V> supplier) {
+        final StableValue<V> stable = StableValue.of();
+
+        final class BgRunnable implements Runnable {
+
+            volatile Thread thread;
+
+            @Override
+            public void run() {
+                try {
+                    stable.computeIfUnset(supplier);
+                } catch (Throwable throwable) {
+                    final Thread.UncaughtExceptionHandler uncaughtExceptionHandler =
```

Comment:
Does this exception handling differ from the default one for threads? If not, I think we can safely remove this catch block, as all exceptions are just propagated and computeIfUnset doesn't declare any checked exception.

Reply:
Nice catch. This will reduce complexity:

```
            @Override
            public void run() {
                stable.computeIfUnset(supplier);
                // Exceptions are implicitly captured by the tread's
                // uncaught exception handler.
            }
```